pkg/hive/hive.go

```diff
@@ -0,0 +1,126 @@
+	"time"
+
+	"github.com/ethersphere/bee/pkg/addressbook"
+	"github.com/ethersphere/bee/pkg/discovery"
+	"github.com/ethersphere/bee/pkg/hive/pb"
+	"github.com/ethersphere/bee/pkg/logging"
+	"github.com/ethersphere/bee/pkg/p2p"
+	"github.com/ethersphere/bee/pkg/p2p/protobuf"
+	"github.com/ethersphere/bee/pkg/swarm"
+
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+const (
+	protocolName    = "hive"
+	protocolVersion = "1.0.0"
+	peersStreamName = "peers"
+	messageTimeout  = 5 * time.Second // maximum allowed time for a message to be read or written.
+	maxBatchSize    = 50
+)
+
+type Service struct {
+	streamer    p2p.Streamer
+	addressBook addressbook.GetterPutter
+	logger      logging.Logger
+}
+
+type Options struct {
+	Streamer    p2p.Streamer
+	AddressBook addressbook.GetterPutter
+	Logger      logging.Logger
+}
+
+func New(o Options) *Service {
+	return &Service{
+		streamer:    o.Streamer,
+		logger:      o.Logger,
+		addressBook: o.AddressBook,
+	}
+}
+
+func (s *Service) Protocol() p2p.ProtocolSpec {
+	return p2p.ProtocolSpec{
+		Name:    protocolName,
+		Version: protocolVersion,
+		StreamSpecs: []p2p.StreamSpec{
+			{
+				Name:    peersStreamName,
+				Handler: s.peersHandler,
+			},
+		},
+	}
+}
+
+func (s *Service) BroadcastPeers(ctx context.Context, addressee swarm.Address, peers ...discovery.BroadcastRecord) error {
+	for len(peers) > maxBatchSize {
+		if err := s.sendPeers(ctx, addressee, peers[0:maxBatchSize]); err != nil {
+			return err
+		}
+
+		peers = peers[maxBatchSize:]
+	}
+
+	if len(peers) > 0 {
+		if err := s.sendPeers(ctx, addressee, peers); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (s *Service) sendPeers(ctx context.Context, peer swarm.Address, peers []discovery.BroadcastRecord) error {
+	stream, err := s.streamer.NewStream(ctx, peer, protocolName, protocolVersion, peersStreamName)
+	if err != nil {
+		return fmt.Errorf("new stream: %w", err)
+	}
+
+	defer stream.Close()
+
+	w, _ := protobuf.NewWriterAndReader(stream)
+	var peersRequest pb.Peers
+	for _, p := range peers {
+		peersRequest.Peers = append(peersRequest.Peers, &pb.BzzAddress{
+			Overlay:  p.Overlay.Bytes(),
+			Underlay: p.Addr.String(),
+		})
+	}
+
+	if err := w.WriteMsg(&peersRequest); err != nil {
+		return fmt.Errorf("write Peers message: %w", err)
+	}
+
+	return stream.FullClose()
+}
+
+func (s *Service) peersHandler(peer p2p.Peer, stream p2p.Stream) error {
+	defer stream.Close()
+	_, r := protobuf.NewWriterAndReader(stream)
+	var peersReq pb.Peers
+	if err := r.ReadMsgWithTimeout(messageTimeout, &peersReq); err != nil {
+		return fmt.Errorf("read requestPeers message: %w", err)
+	}
+
+	for _, newPeer := range peersReq.Peers {
+		addr, err := ma.NewMultiaddr(newPeer.Underlay)
+		if err != nil {
+			s.logger.Infof("Skipping peer in response %s: %w", newPeer, err)
+			continue
+		}
+
+		// todo: this might be changed depending on where do we decide to connect peers
+		s.addressBook.Put(swarm.NewAddress(newPeer.Overlay), addr)
```

Comment:
I think that we agreed on meetings that if we would connect each time that it would not be long-term, but maybe good only for MVP. Plus, we talked about deciding which peers  to connect to and which peers to  to recommend to other peers. I think, in order to acomplish this we would need to inform some structure/service (topology driver, or address book, or something else) that would make a decision and also keep track of connected and known peers. The first, mvp, simple implementation of this service could be to just connect to each peer, but I don't think that connecting from hive is a good idea. @janos, @acud let me know what you think.

Reply:
@pradovic and I thought we agreed that we move towards a minimal implementation that's supposed to be deployed on a cluster. This iteration does not fulfill this expectation, and I thought we agreed that the topology driver will be able to decide which peer to connect to. It must not be through a `Connect` method (whose semantics necessarily suggests that a connection the requested peer would be made), but maybe through an `AddPeer` method that would subsequently _maybe_ connect to the peer then add the peer to the address book.

I'm not sure what do you mean by developing a "king of connection tracker" - that is the topology driver in itself, so I'm not sure about introducing more synthetic abstractions here. I would definitely not want another moving part in creating a topology (i.e. considering and testing kademlia-to-come implementation which is complex enough with another cog that is some independent component that decides whether to connect to peers according to some other intricate unbeknownst logic - there's no need for semantic differentiation between the "connection tracker" and the topology driver).

Reply:
I agree it can be done in topology driver, as I wrote in the comment. Some kind of connection tracker is just a phrase here (no kings mentioned anywhere), my point was just not to use a `Connect` method directly. So, adding the `AddPeer` method to a topology driver sounds right to me.
Another point is that I thought that address book should contain addresses of known but not connected peers as well, so they can be used to try new connections when number of connections drops. That is why I thought that we should add peer to an address book in each case, although I do agree it could be done by topology driver as well.

Reply:
Also current implementation of the full connectivity topology driver `AddPeer` is called from within `libp2p.Connect` method, and assumes that the address is already in the address book. It seemed to me that it's point was to broadcast the nodes, other then connect or/and add addresses to the address book. That is why I was leaning to another abstraction as I did not want to change the implementation too much.

Reply:
Pls check the last commit and lmnwyt.